sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -172,60 +172,47 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
 }
 
 /**
- * Returns a Row containing the evaluation of all children expressions.
- */
-@ExpressionDescription(
-  usage = "_FUNC_(col1, col2, col3, ...) - Creates a struct with the given field values.")
-case class CreateStruct(children: Seq[Expression]) extends Expression {
-
-  override def foldable: Boolean = children.forall(_.foldable)
-
-  override lazy val dataType: StructType = {
-    val fields = children.zipWithIndex.map { case (child, idx) =>
+* a helper mixin to encapsulate [[CreateStruct]] => [[CreateNamedStruct]] transformation.
+*/
+sealed trait CreateStructLikeFactory[T <: Expression] extends (Seq[Expression] => T) {
```

Comment:
This IMO is way to complicated. We are defining two classes for something which can be easily done with 1 shared method.